x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/NativeMemoryCalculator.java

```diff
@@ -179,4 +188,98 @@ public static long dynamicallyCalculateJvmSizeFromNativeMemorySize(long nativeMa
+                    maxMlNodeSize,
+                    clusterSettings.get(MAX_MACHINE_MEMORY_PERCENT),
+                    clusterSettings.get(USE_AUTO_MACHINE_MEMORY_PERCENT)
+                )
+            );
+        }
+
+        if (maxMlMemory <= 0) {
+            // This implies there are currently no ML nodes in the cluster, and
+            // no automatic mechanism for adding one, so we have no idea what
+            // the effective limit would be if one were added
+            return null;
+        }
+
+        maxMlMemory -= Math.max(Job.PROCESS_MEMORY_OVERHEAD.getBytes(), DataFrameAnalyticsConfig.PROCESS_MEMORY_OVERHEAD.getBytes());
+        maxMlMemory -= MachineLearning.NATIVE_EXECUTABLE_CODE_OVERHEAD.getBytes();
+        return ByteSizeValue.ofMb(ByteSizeUnit.BYTES.toMB(Math.max(0L, maxMlMemory)));
+    }
+
+    public static ByteSizeValue calculateTotalMlMemory(ClusterSettings clusterSettings, DiscoveryNodes nodes) {
+
+        long totalMlMemory = 0;
+
+        for (DiscoveryNode node : nodes) {
+            OptionalLong limit = allowedBytesForMl(node, clusterSettings);
+            if (limit.isEmpty()) {
+                continue;
+            }
+            totalMlMemory += limit.getAsLong();
+        }
+
+        // Round down to a whole number of megabytes, since we generally deal with model
+        // memory limits in whole megabytes
+        return ByteSizeValue.ofMb(ByteSizeUnit.BYTES.toMB(totalMlMemory));
+    }
+
+    /**
+     * Get the maximum value of model memory limit that a user may set in a job config.
+     * If the xpack.ml.max_model_memory_limit setting is set then the value comes from that.
+     * Otherwise, if xpack.ml.use_auto_machine_memory_percent is set then the maximum model
+     * memory limit is considered to be the largest model memory limit that could fit into
+     * the cluster (on the assumption that configured lazy nodes will be added and other
+     * jobs stopped to make space).
+     * @return The maximum model memory limit calculated from the current cluster settings,
+     *         or {@link ByteSizeValue#ZERO} if there is no limit.
+     */
+    public static ByteSizeValue getMaxModelMemoryLimit(ClusterService clusterService) {
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        ByteSizeValue maxModelMemoryLimit = clusterSettings.get(MachineLearningField.MAX_MODEL_MEMORY_LIMIT);
+        if (maxModelMemoryLimit != null && maxModelMemoryLimit.getBytes() > 0) {
+            return maxModelMemoryLimit;
+        }
+        // When the ML memory percent is being set automatically and no explicit max model memory limit is set,
+        // max model memory limit is considered to be the max model memory limit that will fit in the cluster
+        Boolean autoMemory = clusterSettings.get(MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT);
+        if (autoMemory) {
+            DiscoveryNodes nodes = clusterService.state().getNodes();
```

Comment:
I initially thought we should filter these on ones where job allocation is enabled, then I saw we are taking advantage of the side-effect of not setting memory limits on the node unless it has the `ml` role. 

This seems possibly unsafe. It would be nice to filter by role here to be explicit and pass in the list of nodes. 

What do you think?

Reply:
Doing this smoked out a few inconsistencies in our existing unit tests.  All the fallout from this is in the second commit.